pysph/base/tree/point_octree.mako

```diff
@@ -0,0 +1,367 @@
+        % else:
+            pid_dst = pids_dst[pbound_here.s0 + lid];
+        % endif
+
+        xd = xdst[pid_dst];
+        yd = ydst[pid_dst];
+        zd = zdst[pid_dst];
+        hd = hdst[pid_dst];
+    }
+
+    // Set loop parameters
+    int cid_src, pid_src;
+    int offset_src = neighbor_cid_offset[idx];
+    int offset_lim = neighbor_cid_offset[idx + 1];
+    uint2 pbound_here2;
+    int m;
+    local ${data_t} xs[${wgs}];
+    local ${data_t} ys[${wgs}];
+    local ${data_t} zs[${wgs}];
+    local ${data_t} hs[${wgs}];
+    ${data_t} r2;
+
+
+    ${caller.pre_loop()}
+    while (offset_src < offset_lim) {
+        cid_src = neighbor_cids[offset_src];
+        pbound_here2 = pbounds_src[cid_src];
+
+        offset_src++;
+
+        while (pbound_here2.s0 < pbound_here2.s1) {
+            // Copy src data
+            if (pbound_here2.s0 + lid < pbound_here2.s1) {
+                %if sorted:
+                    pid_src = pbound_here2.s0 + lid;
+                % else:
+                    pid_src = pids_src[pbound_here2.s0 + lid];
+                %endif
+                xs[lid] = xsrc[pid_src];
+                ys[lid] = ysrc[pid_src];
+                zs[lid] = zsrc[pid_src];
+                hs[lid] = hsrc[pid_src];
+            }
+            m = min(pbound_here2.s1, pbound_here2.s0 + ${wgs}) - pbound_here2.s0;
+
+            barrier(CLK_LOCAL_MEM_FENCE);
+
+            // Everything this point forward is done independently
+            // by each thread.
+            if (svalid) {
+                for (int j=0; j < m; j++) {
+                    % if sorted:
+                        pid_src= pbound_here2.s0 + j;
+                    % else:
+                        pid_src = pids_src[pbound_here2.s0 + j];
+                    % endif
+                    ${data_t} dist2 = NORM2(xs[j] - xd,
+                                            ys[j] - yd,
+                                            zs[j] - zd);
+
+                    r2 = MAX(hs[j], hd) * radius_scale;
+                    r2 *= r2;
+                    if (dist2 < r2) {
+                        ${caller.query()}
+                    }
+                }
+            }
+            pbound_here2.s0 += ${wgs};
+            barrier(CLK_LOCAL_MEM_FENCE);
+        }
+    }
+    ${caller.post_loop()}
+</%def>
+
+<%def name="find_neighbor_counts_args(data_t, sorted, wgs)" cached="False">
+    int *unique_cids, int *pids_src, int *pids_dst, int *cids,
+    uint2 *pbounds_src, uint2 *pbounds_dst,
+    ${data_t} *xsrc, ${data_t} *ysrc, ${data_t} *zsrc, ${data_t} *hsrc,
+    ${data_t} *xdst, ${data_t} *ydst, ${data_t} *zdst, ${data_t} *hdst,
+    ${data_t} radius_scale,
+    int *neighbor_cid_offset, int *neighbor_cids,
+    int *neighbor_counts
+</%def>
+<%def name="find_neighbor_counts_src(data_t, sorted, wgs)" cached="False">
+     <%self:find_neighbors_template data_t="${data_t}" sorted="${sorted}" wgs="${wgs}">
+        <%def name="pre_loop()">
```

Comment:
The nested def syntax is hard to read unfortunately.  Actually the entire template we have is pretty hard to read as it really does not look very "templatey".  I would like this to be much more readable.